src/plugins/dashboard/public/application/lib/build_dashboard_container.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import _ from 'lodash';
+import { DashboardSavedObject } from '../../saved_dashboards';
+import { DashboardContainer, DASHBOARD_CONTAINER_TYPE } from '../embeddable';
+import {
+  DashboardBuildContext,
+  DashboardState,
+  DashboardContainerInput,
+  DashboardAppServices,
+} from '../../types';
+import {
+  enableDashboardSearchSessions,
+  getSearchSessionIdFromURL,
+  stateToDashboardContainerInput,
+} from '.';
+import {
+  ContainerOutput,
+  EmbeddableFactoryNotFoundError,
+  EmbeddableInput,
+  EmbeddablePackageState,
+  ErrorEmbeddable,
+  isErrorEmbeddable,
+} from '../../services/embeddable';
+
+type BuildDashboardContainerProps = DashboardBuildContext & {
+  data: DashboardAppServices['data']; // the whole data service is required here because it is required by getUrlGeneratorState
+  savedDashboard: DashboardSavedObject;
+  initialDashboardState: DashboardState;
+  incomingEmbeddable?: EmbeddablePackageState;
+};
+
+/**
+ * Builds the dashboard container and manages initial search session
+ */
+export const buildDashboardContainer = async ({
+  getLatestDashboardState,
+  initialDashboardState,
+  isEmbeddedExternally,
+  dashboardCapabilities,
+  incomingEmbeddable,
+  savedDashboard,
+  kibanaVersion,
+  embeddable,
+  history,
+  data,
+}: BuildDashboardContainerProps) => {
+  const {
+    search: { session },
+  } = data;
+
+  // set up search session
+  enableDashboardSearchSessions({
+    data,
+    kibanaVersion,
+    savedDashboard,
+    initialDashboardState,
+    getLatestDashboardState,
+    canStoreSearchSession: dashboardCapabilities.storeSearchSession,
+  });
+  const searchSessionIdFromURL = getSearchSessionIdFromURL(history);
+  if (searchSessionIdFromURL) {
+    session.restore(searchSessionIdFromURL);
+  }
+
+  const dashboardFactory = embeddable.getEmbeddableFactory<
+    DashboardContainerInput,
+    ContainerOutput,
+    DashboardContainer
+  >(DASHBOARD_CONTAINER_TYPE);
+
+  if (!dashboardFactory) {
+    throw new EmbeddableFactoryNotFoundError('dashboard app requires dashboard embeddable factory');
+  }
+
+  // Build the initial input for the dashboard container based on the dashboard state.
+  const initialInput = stateToDashboardContainerInput({
+    searchSessionId: searchSessionIdFromURL ?? session.start(),
+    isEmbeddedExternally: Boolean(isEmbeddedExternally),
+    dashboardState: initialDashboardState,
+    dashboardCapabilities,
+    incomingEmbeddable,
+    query: data.query,
+    savedDashboard,
+  });
+
+  /**
+   * Handle the Incoming Embeddable Part 1:
```

Comment:
Added some more thorough comments explaining how the incoming embeddable is added to the dashboard. This can be quite confusing.